### ISSUE TITLE
STT0078: Train whisper model

### ISSUE THREAD

**Issue by gangagyatso4364** [spsither/stt-whisper#3]
### Description 
we are going to get an estimation of time and cost for training the whisper model. the whisper model is previously trained on 500 hours of training data but now we have 1500 hours training data so there will be significant improvement in this model compared to last time. 

[whisper paper](https://cdn.openai.com/papers/whisper.pdf)
[whisper base model code](https://github.com/openai/whisper)
[whisper  paper vedio](https://www.youtube.com/watch?v=AwJf8aQfChE&t=1825s)

### Completion Criteria 
A trained whisper model on 1500 hours of training data
### Implementaion
1. set up the ssh server instance with l4 24 gpu 
2. convert the finetune-whisper notebook to pyhton file
3. set up tmux session in the server
4. train the model with proper training arguments 
5. save the trained model to hf
6. run the prevent hallucination script
7. evaluate the model 
8. compare with previous whisper model
### Subtask 
- [x] set ssh server with neccessary specs
- [x] conversion of finetuning script to python file from jupyter notebook
- [x] run the training in tmux session
- [x] save the models in hf
- [x] run prevent hallucination script 
- [x] run inference on benchmark with finetuned model checkpoint 60k(lowest wer) and 95k(latest). 
- [x] evaluate the model
- [x] compare with previous models
- [x] create a documentation of models comparison

**Comment by gangagyatso4364** [comment]
have created a GPU allocation card request in asana board. 

whisper small training on 1500 hours latest training data with new departments compared to 500 hours previously trained limited data variation in training. So we expect a better wer and cer score this time with more training data and variations in training data. 

time and spec:
epoch: 10
training data hours: 1500
4 (L4) 24 Gpu for 10 days or
8 (L4) 24 Gpu for 10 days

**Comment by gangagyatso4364** [comment]
model stopped training at 4.7 epoch as no improvements in val wer shown.

**Comment by kaldan007** [comment]
@TenzinGayche can u share the pipeline link here

**Comment by gangagyatso4364** [comment]
wand report: https://wandb.ai/TMLM/stt_whisper/runs/zplrw5l5?nw=nwuserganga2000
train:
![Image](https://github.com/user-attachments/assets/7e482823-22e3-404e-ab40-742d932898cb)

**Comment by gangagyatso4364** [comment]
eval:
![Image](https://github.com/user-attachments/assets/42f9d1cc-77e4-4fba-8f2c-f6a3ac05bff7)

**Comment by gangagyatso4364** [comment]
```
model.generation_config.no_repeat_ngram_size = 2
model.generation_config.length_penalty = -1.0
model.generation_config.num_beams = 3   
```
These configuration settings for the Whisper model adjust its text generation behavior:

## no_repeat_ngram_size = 2

This setting prevents the model from repeating any 2-gram (sequence of two words) in the generated output[1]. It helps reduce repetitive hallucinations, which Whisper models can be prone to[2].

## length_penalty = -1.0

The length penalty influences the model's preference for shorter or longer sequences. A negative value (-1.0) encourages the model to generate shorter sequences.

## num_beams = 3

This activates beam search decoding with 3 beams[1][3]. Beam search considers multiple possible sequence continuations at each step, potentially leading to higher quality output compared to greedy search. It helps the model explore different possibilities and choose the overall best sequence[3].

These settings combined aim to:
1. Reduce repetition in the transcription
2. Favor concise outputs
3. Improve overall transcription quality by considering multiple sequence possibilities

**Comment by gangagyatso4364** [comment]
---

### Benchmark Inference Overview
- **Dataset**: 62,000 data points.
- **Duration**: Inference is expected to take ~3 days.

### Training Status
- **Training Stopped**: Validation loss and WER started increasing slightly after checkpoint 60k.

### Evaluation Plan
- **Checkpoints**:
  - **60k Checkpoint**: Lowest WER and validation loss so far.
  - **95k Checkpoint**: Latest checkpoint, includes all training updates.
- **Objective**: Compare CER (Character Error Rate) between checkpoints.
  - If **95k CER < 60k CER**, the latest checkpoint is better despite slightly higher WER.

### Next Steps
- **Resume Fine-tuning**: Continue training if 95k performs better, focusing on improving CER while monitoring validation metrics.

---

**Comment by gangagyatso4364** [comment]
### Importance of Evaluating CER and WER on Tibetan Unicode

1. **Accuracy**: 
   - Tibetan Unicode directly represents the actual language and script, capturing script-specific features that Wylie transliteration cannot.

2. **Linguistic Relevance**: 
   - Metrics computed on Tibetan script are more meaningful for native speakers and linguists, providing insights aligned with real-world language use.

3. **Consistency**: 
   - Unicode eliminates ambiguities and inconsistencies in representation that can arise with Wylie transliteration, ensuring reliable evaluations.

4. **Pronunciation Assessment**: 
   - Evaluating on Tibetan script allows for better detection of pronunciation errors, which may be masked or misrepresented in Wylie.

5. **Modern Standards**: 
   - Tibetan Unicode aligns with current best practices in natural language processing and digital linguistics, making it the preferred choice for modern applications.

---

Switching to Tibetan Unicode for evaluation ensures a more accurate, consistent, and linguistically appropriate assessment of the speech recognition model's performance.

**Comment by gangagyatso4364** [comment]
**Before wylie compute:**
```
   
def compute_metrics(pred):
    pred_ids = pred.predictions
    label_ids = pred.label_ids

    # replace -100 with the pad_token_id
    label_ids[label_ids == -100] = tokenizer.pad_token_id

    # we do not want to group tokens when computing the metrics
    pred_str = tokenizer.batch_decode(pred_ids, skip_special_tokens=True)
    label_str = tokenizer.batch_decode(label_ids, skip_special_tokens=True)

    wer = 100 * min(metric.compute(predictions=pred_str, references=label_str), 1)

    return {"wer": wer}
```

**After tibetan text compute:**
```
def compute_metrics(pred):
    pred_ids = pred.predictions
    label_ids = pred.label_ids

    # Replace -100 with the pad_token_id
    label_ids[label_ids == -100] = tokenizer.pad_token_id

    # Decode predictions and labels from IDs
    pred_str = tokenizer.batch_decode(pred_ids, skip_special_tokens=True)
    label_str = tokenizer.batch_decode(label_ids, skip_special_tokens=True)

    # Convert Wylie to Tibetan Unicode
    pred_unicode = [converter.toUnicode(text) for text in pred_str]
    label_unicode = [converter.toUnicode(text) for text in label_str]

    # Split Tibetan Unicode text into words using tsek (་) as the separator
    pred_words = [text.split("་") for text in pred_unicode]
    label_words = [text.split("་") for text in label_unicode]

    # Compute CER (Character Error Rate) directly on Unicode strings
    cer = 100 * metric.compute(predictions=pred_unicode, references=label_unicode, rouge_type="cer")

    # Compute WER (Word Error Rate) using tsek-based word splitting
    wer = 100 * metric.compute(predictions=pred_words, references=label_words, rouge_type="wer")

    return {"cer": cer, "wer": wer}
```

**Comment by kaldan007** [comment]
If the latest check point is better u can fine tune the model with high quality data for one epoc @gangagyatso4364

**Comment by gangagyatso4364** [comment]
| Model Name          | Model Type | Benchmark mean CER | STT_AB | STT_CS | STT_MV | STT_NS | STT_TT | STT_HS | STT_NW | STT_PC | BENCHMARK |
|---------------------|------------|--------------------|--------|--------|--------|--------|--------|--------|--------|--------|-----------|
| whisper-small-60k        | whisper    | 19.75%             | 10.75% | 26.95% | 26.35% | 25.27% | 16.53% | 25.06% | 11.06% | 16.04% |  [benchmark v3](https://huggingface.co/datasets/ganga4364/stt_benchmark)         |
| whisper-small-95k        | whisper    | 19.19%             | 10.80% | 25.92% | 25.91% | 24.54% | 15.37% | 24.38% | 10.78% | 15.81% |  [ benchmark v3](https://huggingface.co/datasets/ganga4364/stt_benchmark)         |         |
| whisper-small-60k        | whisper    | 21.71%             | 6.05%  | 41.74% | 22.00% | 27.22% | 14.03% |        |        |        |  [benchmark v1](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark/resolve/efc6be7d280c458cbad16df58428ae4378706865/test.csv)        |         |
| whisper-small-95k        | whisper    | 21.24%             | 5.96%  | 40.84% | 21.73% | 27.30% | 14.13% |        |        |        | [benchmark v1](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark/resolve/efc6be7d280c458cbad16df58428ae4378706865/test.csv)        |          |
| ganga/mms_300_v4                 | wav2vec2   |15.23%      | 8.41%  | 29.89% | 19.34% | 10.27% | 13.24% | 19.28% | 11.18% | 10.23%|  [benchmark v2](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark)     |
| wadwani/wav2vec2               | mms_300  |  13.22% |  5.11%  |  29.79% | 15.18%  | 10.48%  | 9.86 % | 18.65% | 9.2%   | 7.51 % |  custom benchmark        |
| spsither/mms_300_v3.1020*       | mms_300  | 18.16% | 4.05% | 33.11% | 18.03% | 21.70% |  8.39% | 16.58%  |  7.49%   |  6.3%   |  [benchmark v1](https://huggingface.co/datasets/openpecha/tibetan-voice-benchmark/resolve/efc6be7d280c458cbad16df58428ae4378706865/test.csv)        |
